src/lib/core/strftime.c

```diff
@@ -0,0 +1,685 @@
+							w = 1;
+							break;
+						}
+						if (yday >= bot) {
+							w = 1 + ((yday - bot) /
+								 DAYSPERWEEK);
+							break;
+						}
+						--base;
+						yday += isleap_sum(year, base)
+						    ? DAYSPERLYEAR
+						    : DAYSPERNYEAR;
+					}
+#ifdef XPG4_1994_04_09
+					if ((w == 52 &&
+					     t->tm_mon == TM_JANUARY) ||
+					    (w == 1 &&
+					     t->tm_mon == TM_DECEMBER))
+						w = 53;
+#endif				/* defined XPG4_1994_04_09 */
+					if (*format == 'V')
+						pt = _conv(w, "%02d",
+							   pt, ptlim);
+					else if (*format == 'g') {
+						*warnp = IN_ALL;
+						pt = _yconv(year, base,
+							    false, true,
+							    pt, ptlim);
+					} else
+						pt = _yconv(year, base,
+							    true, true,
+							    pt, ptlim);
+				}
+				continue;
+			case 'v':
+				/*
+				 ** From Arnold Robbins' strftime version 3.0:
+				 ** "date as dd-bbb-YYYY"
+				 ** (ado, 1993-05-24)
+				 */
+				pt = _fmt("%e-%b-%Y", t, pt, ptlim, warnp);
+				continue;
+			case 'W':
+				pt = _conv((t->tm_yday + DAYSPERWEEK -
+					    (t->tm_wday ? (t->tm_wday - 1)
+					     : (DAYSPERWEEK -
+						1))) /
+					   DAYSPERWEEK, "%02d", pt, ptlim);
+				continue;
+			case 'w':
+				pt = _conv(t->tm_wday, "%d", pt, ptlim);
+				continue;
+			case 'X':
+				pt = _fmt(Locale->X_fmt, t, pt, ptlim, warnp);
+				continue;
+			case 'x':
+				{
+					enum warn warn2 = IN_SOME;
+
+					pt = _fmt(Locale->x_fmt, t, pt, ptlim,
+						  &warn2);
+					if (warn2 == IN_ALL)
+						warn2 = IN_THIS;
+					if (warn2 > *warnp)
+						*warnp = warn2;
+				}
+				continue;
+			case 'y':
+				*warnp = IN_ALL;
+				pt = _yconv(t->tm_year, TM_YEAR_BASE,
+					    false, true, pt, ptlim);
+				continue;
+			case 'Y':
+				pt = _yconv(t->tm_year, TM_YEAR_BASE,
+					    true, true, pt, ptlim);
+				continue;
+			case 'Z':
+				if (t->tm_isdst >= 0)
+					pt = _add(tzname[t->tm_isdst != 0],
+						  pt, ptlim);
+				/*
+				 ** C99 and later say that %Z must be
+				 ** replaced by the empty string if the
+				 ** time zone abbreviation is not
+				 ** determinable.
+				 */
+				continue;
+			case 'z':
+				{
+					long diff;
+					char const *sign;
+					bool negative;
+
+					diff = t->tm_gmtoff;
+					negative = diff < 0;
+					if (diff == 0) {
+						negative = t->tm_isdst < 0;
+						if (tzname[t->tm_isdst != 0][0]
+						    == '-')
+							negative = true;
+					}
+					if (negative) {
+						sign = "-";
+						diff = -diff;
+					} else
+						sign = "+";
+					pt = _add(sign, pt, ptlim);
+					diff /= SECSPERMIN;
+					diff = (diff / MINSPERHOUR) * 100 +
+					    (diff % MINSPERHOUR);
+					pt = _conv(diff, "%04d", pt, ptlim);
+				}
+				continue;
+			case '+':
+				pt = _fmt(Locale->date_fmt, t, pt, ptlim,
+					  warnp);
+				continue;
+			case '%':
+				/*
+				 ** X311J/88-090 (4.12.3.5): if conversion char is
+				 ** undefined, behavior is undefined. Print out the
+				 ** character itself as printf(3) also does.
+				 */
+			default:
```

Comment:
It seems that if You write `%<something unknown>` it will pass these symbols to result string.

So You could call `strftime` and then add your own `%<something>`.

But the variant is faster, of course.

Reply:
Yup, it would be double scan of a various tring buffers. (G)LIBC `strftime` would process input format string, generate output, and we would process their output string, generating _yet another_ result. And we would need 2 temporary buffers on the fly.

The simpler approach (which is used by all projects I saw) is just to copy Olson' code which is in public domain, and adapt to their needs (BSD llibc, Android, PostgreSQL - all do this for their own needs). See no reason to not use the similar approach here. Such modification allows to avoid information lost in between worlds, like it would happen here if we try to convert `.tzoffset`/`.tzindex` here to `struct tm` and libc `tzname`, and then reconstruct original information from that it's there in [g]libc. When we work directly with `struct datetime` we could entirely avoid any kinds of information loss here, and be libc agnostic. [See '%z', '%f']